panther_hardware_interfaces/test/panther_system/gpio/test_gpio_controller.cpp

```diff
@@ -57,118 +64,129 @@ class TestGPIOController : public ::testing::Test
 protected:
   float GetRobotVersion();
 
+  std::shared_ptr<MockGPIODriver> gpio_driver_;
   std::unique_ptr<GPIOControllerWrapper> gpio_controller_wrapper_;
   static constexpr int watchdog_edges_per_100ms_ = 10;
 };
 
 TestGPIOController::TestGPIOController()
 {
-  if (GetRobotVersion() < 1.2 - std::numeric_limits<float>::epsilon()) {
-    throw std::runtime_error("Tests for this robot versions are not implemented");
-  }
+  gpio_driver_ = std::make_shared<MockGPIODriver>();
 
-  gpio_controller_wrapper_ = std::make_unique<GPIOControllerWrapper>();
+  // Mock methods called during the initialization process
+  ON_CALL(*gpio_driver_, SetPinValue(GPIOPin::VMOT_ON, true)).WillByDefault(testing::Return(true));
+  ON_CALL(*gpio_driver_, SetPinValue(GPIOPin::DRIVER_EN, true))
+    .WillByDefault(testing::Return(true));
+  ON_CALL(*gpio_driver_, IsPinAvailable(GPIOPin::WATCHDOG)).WillByDefault(testing::Return(true));
+
+  gpio_controller_wrapper_ = std::make_unique<GPIOControllerWrapper>(gpio_driver_);
   gpio_controller_wrapper_->Start();
 }
 
-float TestGPIOController::GetRobotVersion()
+struct GPIOTestParam
 {
-  const char * robot_version_env = std::getenv("PANTHER_ROBOT_VERSION");
-
-  if (!robot_version_env) {
-    throw std::runtime_error("Can't read 'PANTHER_ROBOT_VERSION' environment variable");
-  }
-
-  return std::stof(robot_version_env);
-}
+  GPIOPin pin;
+  std::function<bool(GPIOControllerWrapper *, bool)> enable_method;
+  bool is_inverted = false;
+};
 
-TEST_F(TestGPIOController, TestMotorsInit)
+class ParametrizedTestGPIOController : public TestGPIOController,
+                                       public ::testing::WithParamInterface<GPIOTestParam>
 {
-  EXPECT_TRUE(gpio_controller_wrapper_->IsPinActive(GPIOPin::VMOT_ON));
-  EXPECT_TRUE(gpio_controller_wrapper_->IsPinActive(GPIOPin::DRIVER_EN));
-}
+};
 
-TEST_F(TestGPIOController, TestWatchdog)
+TEST(TestGPIOControllerInitialization, GPIODriverUninitialized)
 {
-  auto edge_cnt = std::make_shared<int>(0);
+  std::shared_ptr<MockGPIODriver> gpio_driver;
 
-  gpio_controller_wrapper_->RegisterGPIOEventCallback([this, edge_cnt](const auto & state) mutable {
-    if (state.pin == GPIOPin::WATCHDOG) {
-      (*edge_cnt)++;
-    }
-  });
+  EXPECT_THROW(std::make_unique<GPIOControllerWrapper>(gpio_driver), std::runtime_error);
+}
 
-  gpio_controller_wrapper_->WatchdogEnable();
-  ASSERT_TRUE(gpio_controller_wrapper_->IsWatchdogEnabled());
+TEST(TestGPIOControllerInitialization, WatchdogPinNotAvailable)
+{
+  auto gpio_driver = std::make_shared<MockGPIODriver>();
 
-  std::this_thread::sleep_for(std::chrono::milliseconds(100));
-  gpio_controller_wrapper_->WatchdogDisable();
+  EXPECT_CALL(*gpio_driver, SetPinValue(testing::_, true))
+    .Times(2)
+    .WillRepeatedly(testing::Return(true));
```

Comment:
Why function SetPinValue is checkek?

Reply:
We don't check here. When we operate on mocks we define their behavior after specific methods are called. So here our goal is to tell the mock to return `true` twice when the `SetPinValue` method is invoked with any pin and `true` as a second parameter. The `SetPinValue` is invoked during the initialization of GPIOController (we don't use a test fixture so it's needed to do it manually).

PS. We could use the `ON_CALL` as well but it's known that initialization demands the `SetPinValue` method to be called twice and verify the flow by chance.

Reply:
ok